include/envoy/matcher/matcher.h

```diff
@@ -0,0 +1,121 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/config/common/matcher/v3/matcher.pb.h"
+#include "envoy/config/typed_config.h"
+
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+#include "external/envoy_api/envoy/config/core/v3/extension.pb.h"
+
+namespace Envoy {
+namespace Matcher {
```

Comment:
High level comment: I'm finding it very hard to read this file and understand how this is going to be used. Can you potentially add more comments and if it makes it easier some brief discussion of some example of what a concrete impl would look like? Up to you if you want to do this above each class or as a top level comment, but I think it would really help me with the rest of the review. Thanks.

Reply:
Yeah makes sense, I've added a top level comment that tries to explain it, lmk if that clarifies things.